examples/js/loaders/GLTFLoader.js

```diff
@@ -1571,51 +1639,91 @@ THREE.GLTFLoader = ( function () {
 			switch ( type ) {
 
 				case 'scene':
-					dependency = this.loadScene( index );
+					dependency = this.loadScene( index ).then( function ( scene ) {
+
+						return parser._onAfter( 'Scene', scene, json.scenes[ index ] );
+
+					} );
 					break;
 
 				case 'node':
-					dependency = this.loadNode( index );
+					dependency = this.loadNode( index ).then( function ( node ) {
+
+						return parser._onAfter( 'Node', node, json.nodes[ index ] );
+
+					} );
```

Comment:
Just curious — is it possible to put all the hooks into `onDependency`?  E.g...

```js
case 'node':
  dependency = this._onBefore( type, index )
    .then(( def ) => this._on( type, index, def ))
    .then(( result, def ) => this.load( type, index, def ))
    .then(( result ) => this._onAfter( type, index, def, result ));
  break;
```

That would be simpler to maintain, but I don't know if it's compatible with all the cases we need.

We may also want to measure parsing time before/after here. I'm a little worried that adding a chain of promises for every accessor might create a lot of overhead.

Reply:
> is it possible to put all the hooks into `onDependency`

I haven't looked into the all asset types yet and perhaps we need a bit change in `.loadXXX()`, but probably it's possible. Maybe like this?

```javascript
// .getDependency()
case 'XXX':
  dependency = this._onBefore(XXX, index (or json.XXXs[index])).then(function (def) {
    return parser._loadXXX(index, def, parser._on(XXX, def)).then(function (result) {
      return parser._onAfter(XXX, result, def);
    });
  });
  break;

// .loadXXX()
/**
 * @param {number} xxxIndex
 * @param {GLTF.definition} xxxDef
 * @param {Promise<Object>|null} customXXX (optional)
 * @return {Promise<Object>}
 */
GLTFParser.prototype.loadXXX = function (xxxIndex, xxxDef, customXXX) {
  var pending = customXXX;
  if (!pending) {
    pending = create core spec object from xxxDef (and xxxIndex);
  }
  pending.then(function (xxx) {
    // setting up common properties here
    // for example name, position/quaternion/scale/matrix, and so on for Node
    return xxx;
  } );
};
```

> We may also want to measure parsing time before/after here.

Yes, that's the plan. I first wanted to share the rough sketch API idea.

Reply:
In the above my comment, I thought to pass `customXXX` to `loadXXX()` to let it set up common properties [for example the ones in `loadMaterial()`]( https://github.com/mrdoob/three.js/blob/017d34812c2fd0254f4bbf936b1ef7ac7dba0810/examples/js/loaders/GLTFLoader.js#L2307-L2336) but coming to think that it can be a bit messy and some extension handlers may not want the common properties to be overridden because they set the properties something special.

So considering to call `.loadXXX()` only if `._onXXX()` doesn't create an object. Pros: `.loadXXX()` doesn't need to change other than taking from `index` to `def` (and `index`), we can keep the code clean. Cons: Extension handlers may have duplicated codes with the ones in the parser because they need to even set up the common properties by themselves.

Reply:
With the API above (onBeforeXXX, onXXX, and onAfterXXX for all asset types + Map, Geometry, and GLTF) I measured the loading response time (from `.load()` to `onLoad`) of some models in gltf extension example. The numbers are average of five time trials. I haven't looked into the detail yet but Promise chain seems to have non-negligible response time impact. But I guess (and hope) main thread blocking time may not be so different? If so, I hope this impact is acceptable.

(I'm not sure yet why the Monster glTF-Draco numbers between dev and plugin are close... Five time trial is not may enough so just occasional? Or draco uncompression processing is dominant?)

| | dev [ms] | plugin [ms] | plugin / dev [%] |
| ---- | ---- | ---- | ---- |
| Boombox glTF | 42.81 | 51.13 | +19% |
| Boombox glTF-dds | 205.70 | 225.52 | +10% |
| Monster glTF | 49.06 | 74.62 | +52% |
| Monster glTF-Draco | 167.96 | 164.90 | -2% |